hedera-node/src/main/java/com/hedera/services/store/contracts/HederaWorldState.java

```diff
@@ -407,27 +333,58 @@ public void revert() {
 		@Override
 		public void commit() {
 			final HederaWorldState wrapped = wrappedWorldView();
-			wrapped.provisionalAccountDeletes.addAll(getDeletedAccountAddresses());
-			wrapped.sponsorMap.putAll(sponsorMap);
+			var repository = wrapped.repositoryRoot;
+
+			getDeletedAccountAddresses().forEach(address -> {
+				// account may have been created and deleted within the same transaction.
+				if (repository.isExist(address.toArray())) {
+					var account = repository.getAccount(address.toArray());
+					// selfdestruct adjusts balances to zero
+					account.setBalance(BigInteger.ZERO);
+					validateTrue(!repository.isDeleted(address.toArray()), FAIL_INVALID);
+					account.setDeleted(true);
+				}
+			});
 
 			for (final UpdateTrackingAccount<WorldStateAccount> updated : getUpdatedAccounts()) {
+
 				final WorldStateAccount origin = updated.getWrappedAccount();
+				byte[] address = updated.getAddress().toArray();
+				long balance = updated.getBalance().toLong();
```

Comment:
Instead of converting it to `long` and from `long` to bigInteger, you can do directly `updatedAccount.setBalance(updated.getBalance().toBigInteger())`